File: packages/sdk/go/devbox/webhooks.go
package devbox

import (
	"context"
	"fmt"
	"net/url"
	"time"
)

// WebhooksService handles webhook operations
type WebhooksService struct {
	client *Client
}

// Webhook represents a webhook subscription
type Webhook struct {
	ID          string             `json:"id"`
	Name        string             `json:"name"`
	Description string             `json:"description,omitempty"`
	URL         string             `json:"url"`
	Events      []string           `json:"events"`
	IsActive    bool               `json:"is_active"`
	Headers     map[string]string  `json:"headers,omitempty"`
	RetryConfig *WebhookRetryConfig `json:"retry_config,omitempty"`
	CreatedAt   time.Time          `json:"created_at"`
	UpdatedAt   time.Time          `json:"updated_at"`
}

// WebhookRetryConfig defines retry configuration
type WebhookRetryConfig struct {
	MaxRetries    int     `json:"max_retries"`
	RetryDelayMs  int64   `json:"retry_delay_ms"`
	BackoffFactor float64 `json:"backoff_factor"`
	MaxDelayMs    int64   `json:"max_delay_ms"`
}

// WebhookDelivery represents a webhook delivery attempt
type WebhookDelivery struct {
	ID           string     `json:"id"`
	WebhookID    string     `json:"webhook_id"`
	EventType    string     `json:"event_type"`
	EventID      string     `json:"event_id"`
	Payload      string     `json:"payload"`
	ResponseCode int        `json:"response_code,omitempty"`
	ResponseBody string     `json:"response_body,omitempty"`
	Status       string     `json:"status"`
	AttemptCount int        `json:"attempt_count"`
	NextRetryAt  *time.Time `json:"next_retry_at,omitempty"`
	DeliveredAt  *time.Time `json:"delivered_at,omitempty"`
	ErrorMessage string     `json:"error_message,omitempty"`
	DurationMs   int64      `json:"duration_ms,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
}

// CreateWebhookRequest contains parameters for creating a webhook
type CreateWebhookRequest struct {
	Name        string             `json:"name"`
	Description string             `json:"description,omitempty"`
	URL         string             `json:"url"`
	Events      []string           `json:"events"`
	Headers     map[string]string  `json:"headers,omitempty"`
	RetryConfig *WebhookRetryConfig `json:"retry_config,omitempty"`
}

// WebhookListResponse contains a paginated list of webhooks
type WebhookListResponse struct {
	Data       []Webhook  `json:"data"`
	Pagination Pagination `json:"pagination"`
}

// WebhookDeliveryListResponse contains a paginated list of webhook deliveries
type WebhookDeliveryListResponse struct {
	Data       []WebhookDelivery `json:"data"`
	Pagination Pagination        `json:"pagination"`
}

// List returns a list of webhooks
func (s *WebhooksService) List(ctx context.Context, opts *ListOptions) (*WebhookListResponse, error) {
	path := "/webhooks" + opts.toQuery()
	var resp WebhookListResponse
	if err := s.client.request(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Get returns a webhook by ID
func (s *WebhooksService) Get(ctx context.Context, id string) (*Webhook, error) {
	var webhook Webhook
	if err := s.client.request(ctx, "GET", "/webhooks/"+id, nil, &webhook); err != nil {
		return nil, err
	}
	return &webhook, nil
}

// Create creates a new webhook
func (s *WebhooksService) Create(ctx context.Context, req *CreateWebhookRequest) (*Webhook, error) {
	var webhook Webhook
	if err := s.client.request(ctx, "POST", "/webhooks", req, &webhook); err != nil {
		return nil, err
	}
	return &webhook, nil
}

// Update updates a webhook
func (s *WebhooksService) Update(ctx context.Context, id string, req *CreateWebhookRequest) (*Webhook, error) {
	var webhook Webhook
	if err := s.client.request(ctx, "PUT", "/webhooks/"+id, req, &webhook); err != nil {
		return nil, err
	}
	return &webhook, nil
}

// Delete deletes a webhook
func (s *WebhooksService) Delete(ctx context.Context, id string) error {
	return s.client.request(ctx, "DELETE", "/webhooks/"+id, nil, nil)
}

// SetActive activates or deactivates a webhook
func (s *WebhooksService) SetActive(ctx context.Context, id string, active bool) error {
	req := map[string]bool{"active": active}
	return s.client.request(ctx, "PUT", "/webhooks/"+id+"/active", req, nil)
}

// RegenerateSecret regenerates the webhook secret
func (s *WebhooksService) RegenerateSecret(ctx context.Context, id string) (string, error) {
	var resp struct {
		Secret string `json:"secret"`
	}
	if err := s.client.request(ctx, "POST", "/webhooks/"+id+"/secret", nil, &resp); err != nil {
		return "", err
	}
	return resp.Secret, nil
}

// Test sends a test event to a webhook
func (s *WebhooksService) Test(ctx context.Context, id string) error {
	return s.client.request(ctx, "POST", "/webhooks/"+id+"/test", nil, nil)
}

// ListDeliveries returns webhook delivery history
func (s *WebhooksService) ListDeliveries(ctx context.Context, webhookID string, opts *ListOptions) (*WebhookDeliveryListResponse, error) {
	path := fmt.Sprintf("/webhooks/%s/deliveries%s", webhookID, opts.toQuery())
	var resp WebhookDeliveryListResponse
	if err := s.client.request(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// RetryDelivery retries a failed webhook delivery
func (s *WebhooksService) RetryDelivery(ctx context.Context, webhookID, deliveryID string) error {
	path := fmt.Sprintf("/webhooks/%s/deliveries/%s/retry", webhookID, deliveryID)
	return s.client.request(ctx, "POST", path, nil, nil)
}

// GetAvailableEvents returns all available webhook events
func (s *WebhooksService) GetAvailableEvents(ctx context.Context) ([]string, error) {
	var events []string
	if err := s.client.request(ctx, "GET", "/webhooks/events", nil, &events); err != nil {
		return nil, err
	}
	return events, nil
}

// APIKeysService handles API key operations
type APIKeysService struct {
	client *Client
}

// APIKey represents an API key
type APIKey struct {
	ID          string     `json:"id"`
	Name        string     `json:"name"`
	Description string     `json:"description,omitempty"`
	KeyPrefix   string     `json:"key_prefix"`
	Scopes      []string   `json:"scopes"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
	LastUsedAt  *time.Time `json:"last_used_at,omitempty"`
	LastUsedIP  string     `json:"last_used_ip,omitempty"`
	IsActive    bool       `json:"is_active"`
	CreatedAt   time.Time  `json:"created_at"`
}

// CreateAPIKeyRequest contains parameters for creating an API key
type CreateAPIKeyRequest struct {
	Name        string     `json:"name"`
	Description string     `json:"description,omitempty"`
	Scopes      []string   `json:"scopes"`
	ExpiresAt   *time.Time `json:"expires_at,omitempty"`
}

// CreateAPIKeyResponse contains the created API key with the plain key
type CreateAPIKeyResponse struct {
	APIKey   *APIKey `json:"api_key"`
	PlainKey string  `json:"plain_key"`
}

// UpdateAPIKeyRequest contains parameters for updating an API key
type UpdateAPIKeyRequest struct {
	Name        string   `json:"name,omitempty"`
	Description string   `json:"description,omitempty"`
	Scopes      []string `json:"scopes,omitempty"`
}

// APIKeyListResponse contains a paginated list of API keys
type APIKeyListResponse struct {
	Data       []APIKey   `json:"data"`
	Pagination Pagination `json:"pagination"`
}

// List returns a list of API keys
func (s *APIKeysService) List(ctx context.Context, opts *ListOptions) (*APIKeyListResponse, error) {
	path := "/api-keys" + opts.toQuery()
	var resp APIKeyListResponse
	if err := s.client.request(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Get returns an API key by ID
func (s *APIKeysService) Get(ctx context.Context, id string) (*APIKey, error) {
	var key APIKey
	if err := s.client.request(ctx, "GET", "/api-keys/"+id, nil, &key); err != nil {
		return nil, err
	}
	return &key, nil
}

// Create creates a new API key
func (s *APIKeysService) Create(ctx context.Context, req *CreateAPIKeyRequest) (*CreateAPIKeyResponse, error) {
	var resp CreateAPIKeyResponse
	if err := s.client.request(ctx, "POST", "/api-keys", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Update updates an API key
func (s *APIKeysService) Update(ctx context.Context, id string, req *UpdateAPIKeyRequest) (*APIKey, error) {
	var key APIKey
	if err := s.client.request(ctx, "PUT", "/api-keys/"+id, req, &key); err != nil {
		return nil, err
	}
	return &key, nil
}

// Delete deletes an API key
func (s *APIKeysService) Delete(ctx context.Context, id string) error {
	return s.client.request(ctx, "DELETE", "/api-keys/"+id, nil, nil)
}

// Revoke revokes an API key
func (s *APIKeysService) Revoke(ctx context.Context, id string) error {
	return s.client.request(ctx, "POST", "/api-keys/"+id+"/revoke", nil, nil)
}

// GetAvailableScopes returns all available API key scopes
func (s *APIKeysService) GetAvailableScopes(ctx context.Context) ([]string, error) {
	var scopes []string
	if err := s.client.request(ctx, "GET", "/api-keys/scopes", nil, &scopes); err != nil {
		return nil, err
	}
	return scopes, nil
}

File: packages/sdk/python/clouddevbox/webhooks.py
"""Webhooks Service"""

from dataclasses import dataclass
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import DevBoxClient, AsyncDevBoxClient


@dataclass
class WebhookRetryConfig:
    """Webhook retry configuration"""
    max_retries: int
    retry_delay_ms: int
    backoff_factor: float
    max_delay_ms: int


@dataclass
class Webhook:
    """Webhook subscription"""
    id: str
    name: str
    url: str
    events: list[str]
    is_active: bool
    created_at: str
    updated_at: str
    description: Optional[str] = None
    headers: Optional[dict[str, str]] = None
    retry_config: Optional[WebhookRetryConfig] = None


@dataclass
class WebhookDelivery:
    """Webhook delivery attempt"""
    id: str
    webhook_id: str
    event_type: str
    event_id: str
    payload: str
    status: str
    attempt_count: int
    created_at: str
    response_code: Optional[int] = None
    response_body: Optional[str] = None
    next_retry_at: Optional[str] = None
    delivered_at: Optional[str] = None
    error_message: Optional[str] = None
    duration_ms: Optional[int] = None


class WebhooksService:
    """Webhooks service for sync client"""

    def __init__(self, client: "DevBoxClient"):
        self._client = client

    def list(self, page: int = 1, page_size: int = 20) -> dict:
        """List webhooks"""
        return self._client.request("GET", "/webhooks", params={"page": page, "page_size": page_size})

    def get(self, id: str) -> dict:
        """Get webhook by ID"""
        return self._client.request("GET", f"/webhooks/{id}")

    def create(self, name: str, url: str, events: list[str], **kwargs) -> dict:
        """Create a new webhook"""
        data = {"name": name, "url": url, "events": events, **kwargs}
        return self._client.request("POST", "/webhooks", json=data)

    def update(self, id: str, **kwargs) -> dict:
        """Update a webhook"""
        return self._client.request("PUT", f"/webhooks/{id}", json=kwargs)

    def delete(self, id: str) -> None:
        """Delete a webhook"""
        self._client.request("DELETE", f"/webhooks/{id}")

    def set_active(self, id: str, active: bool) -> None:
        """Set webhook active status"""
        self._client.request("PUT", f"/webhooks/{id}/active", json={"active": active})

    def regenerate_secret(self, id: str) -> str:
        """Regenerate webhook secret"""
        resp = self._client.request("POST", f"/webhooks/{id}/secret")
        return resp["secret"]

    def test(self, id: str) -> None:
        """Send test event to webhook"""
        self._client.request("POST", f"/webhooks/{id}/test")

    def list_deliveries(self, webhook_id: str, page: int = 1, page_size: int = 20) -> dict:
        """List webhook deliveries"""
        return self._client.request(
            "GET", f"/webhooks/{webhook_id}/deliveries",
            params={"page": page, "page_size": page_size}
        )

    def retry_delivery(self, webhook_id: str, delivery_id: str) -> None:
        """Retry a failed delivery"""
        self._client.request("POST", f"/webhooks/{webhook_id}/deliveries/{delivery_id}/retry")

    def get_available_events(self) -> list[str]:
        """Get available webhook events"""
        return self._client.request("GET", "/webhooks/events")


class AsyncWebhooksService:
    """Webhooks service for async client"""

    def __init__(self, client: "AsyncDevBoxClient"):
        self._client = client

    async def list(self, page: int = 1, page_size: int = 20) -> dict:
        """List webhooks"""
        return await self._client.request("GET", "/webhooks", params={"page": page, "page_size": page_size})

    async def get(self, id: str) -> dict:
        """Get webhook by ID"""
        return await self._client.request("GET", f"/webhooks/{id}")

    async def create(self, name: str, url: str, events: list[str], **kwargs) -> dict:
        """Create a new webhook"""
        data = {"name": name, "url": url, "events": events, **kwargs}
        return await self._client.request("POST", "/webhooks", json=data)

    async def update(self, id: str, **kwargs) -> dict:
        """Update a webhook"""
        return await self._client.request("PUT", f"/webhooks/{id}", json=kwargs)

    async def delete(self, id: str) -> None:
        """Delete a webhook"""
        await self._client.request("DELETE", f"/webhooks/{id}")

    async def set_active(self, id: str, active: bool) -> None:
        """Set webhook active status"""
        await self._client.request("PUT", f"/webhooks/{id}/active", json={"active": active})

    async def regenerate_secret(self, id: str) -> str:
        """Regenerate webhook secret"""
        resp = await self._client.request("POST", f"/webhooks/{id}/secret")
        return resp["secret"]

    async def test(self, id: str) -> None:
        """Send test event to webhook"""
        await self._client.request("POST", f"/webhooks/{id}/test")

    async def list_deliveries(self, webhook_id: str, page: int = 1, page_size: int = 20) -> dict:
        """List webhook deliveries"""
        return await self._client.request(
            "GET", f"/webhooks/{webhook_id}/deliveries",
            params={"page": page, "page_size": page_size}
        )

    async def retry_delivery(self, webhook_id: str, delivery_id: str) -> None:
        """Retry a failed delivery"""
        await self._client.request("POST", f"/webhooks/{webhook_id}/deliveries/{delivery_id}/retry")

    async def get_available_events(self) -> list[str]:
        """Get available webhook events"""
        return await self._client.request("GET", "/webhooks/events")
